clang/lib/Tooling/DependencyScanning/DependencyScanningFilesystem.cpp

```diff
@@ -330,3 +353,20 @@ DependencyScanningWorkerFilesystem::openFileForRead(const Twine &Path) {
     return Result.getError();
   return DepScanFile::create(Result.get());
 }
+
+std::error_code DependencyScanningWorkerFilesystem::setCurrentWorkingDirectory(
+    const Twine &Path) {
+  updateWorkingDirForCacheLookup(Path.str());
+  return ProxyFileSystem::setCurrentWorkingDirectory(Path);
+}
+
+void DependencyScanningWorkerFilesystem::updateWorkingDirForCacheLookup(
+    llvm::ErrorOr<std::string> CWD) {
+  if (CWD && !CWD->empty()) {
+    WorkingDirForCacheLookup = *CWD;
+  } else {
+    // The cache lookup functions will not accept relative paths for safety, so
+    // at least make it absolute from a "root".
+    WorkingDirForCacheLookup = "/";
```

Comment:
This could silently use the wrong cache entry; isn't that the same as the current issue, just only when there's a CWD error?

Reply:
This is more of a recovery to avoid assertion hits, assuming the error gets reported.

Reply:
I would prefer if we avoided inventing a dummy value.

What do you think about enforcing that `DependencyScanningWorkerFilesystem` never gets constructed from a VFS that either does not have a working directory, or has one that's relative? We could do that by creating a factory function that returns `llvm::Expected<DependencyScanningWorkerFilesystem>` instead of using the constructor.

For the `setCurrentWorkingDirectory()` call, I think it'd be reasonable to clear our CWD if the new one is empty or relative, return an error code and assume the VFS doesn't get used in that state. If it does, hitting one of the "is file path absolute" assertions is fair game.

Reply:
> For the setCurrentWorkingDirectory() call, I think it'd be reasonable to clear our CWD if the new one is empty or relative, return an error code and assume the VFS doesn't get used in that state. If it does, hitting one of the "is file path absolute" assertions is fair game.

This would be fine for a command-line compiler invocation where you can emit an error and then "abort", but this is undesirable in the context of a libclang client, like an IDE, where some recovery in the presence of errors is useful; for example, clang could still try to create an AST even in the presence of such kind of errors in order to support the editor, instead of providing no functionality.

Reply:
In this case the path will be used when **writing** to the cache though, right? That affects scan results for other TUs just like you are fixing in this PR.

I'm not clear what recovery means here: I think what we want is that the scanner reports an error and stops processing the current TU.  We don't get value from partial scanner results for the current TU like we might for an IDE AST query.

Reply:
> In this case the path will be used when writing to the cache though, right? That affects scan results for other TUs just like you are fixing in this PR.

What is the specific concern? If multiple TUs got errors for working directory and subsequently used the same standard prefix then they may resolve to same path for a relative filename but these TUs are already in an erroneous state already, it's not invalidating their correctness.

> I'm not clear what recovery means here

My general point is that I don't think we should make the assumption that if a VFS implementation emits an error for `setCurrentWorkingDirectory()` then that VFS instance should not ever get used again subsequently.
We could say this is specific for the `DependencyScanningWorkerFilesystem` implementation but I don't see a motivating reason that it should not be able to continue operating instead of crashing if it gets used again.

Reply:
> What is the specific concern? If multiple TUs got errors for working directory and subsequently used the same standard prefix then they may resolve to same path for a relative filename but these TUs are already in an erroneous state already, it's not invalidating their correctness.

The dummy path you're generating can overlap with a real file path.  E.g. failing on "include/foo.h" poisons "/include/foo.h".

> My general point is that I don't think we should make the assumption that if a VFS implementation emits an error for setCurrentWorkingDirectory() then that VFS instance should not ever get used again subsequently.

I think this is fine as long as our cache cannot poison the value for a real path.

Reply:
How is it poisoning the cache? The key for the cache lookup will be "/include/foo.h" and the associated value will be whatever "/include/foo.h" resolves to.

Reply:
Oh wait, I see why it is poisoning, I could change it so that it avoids cache lookups if an error occurred?

Reply:
SGTM

Reply:
See latest commit, I've changed it so it became `llvm::ErrorOr<std::string> WorkingDirForCacheLookup` and `DependencyScanningWorkerFilesystem::getOrCreateFileSystemEntry()` errors out if it needs to use `WorkingDirForCacheLookup` and it's erroneous.

Reply:
That looks good to me.